src/metabase/driver/postgres.clj

```diff
@@ -220,13 +221,15 @@
   [driver [_ arg]]
   (sql.qp/->honeysql driver [:percentile arg 0.5]))
 
+(p/defrecord+ RegexMatchFirst [identifier pattern]
+  hformat/ToSql
+  (to-sql [_]
+    (str "substring(" (hformat/to-sql identifier) " FROM " (hformat/to-sql pattern) ")")))
+
 (defmethod sql.qp/->honeysql [:postgres :regex-match-first]
   [driver [_ arg pattern]]
-  (let [col-name (hformat/to-sql (sql.qp/->honeysql driver arg))]
-    (reify
-      hformat/ToSql
-      (to-sql [_]
-        (str "substring(" col-name " FROM " (hformat/to-sql pattern) ")")))))
+  (let [identifier (sql.qp/->honeysql driver arg)]
+    (->RegexMatchFirst identifier pattern)))
```

Comment:
The main reason for this change is that calling `to-sql` prematurely ended up NOT properly quoting the identifier here in some cases which broke certain queries. This was mixed up a bit; `->honeysql` was supposed to return a HoneySQL form of some sort which is compiled to SQL later. We _were_ doing both in this step. The new record type delays to the `to-sql` compilation stage till later.